### PR TITLE
feat(sync): added tests for sync on demand with the most used upstrea…

### PR DESCRIPTION
**Pull request by aokirisaki**
…m registry services

**What type of PR is this?**

**Which issue does this PR fix**:
#1379 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/blackbox/sync_docker.bats [hunk 1]
```diff
@@ -80,3 +84,125 @@ function teardown_file() {
...
+    [ "$status" -eq 0 ]
+    [ $(echo "${lines[-1]}" | jq '.repositories[0]') = '"archlinux"' ]
+    run curl http://127.0.0.1:8090/v2/archlinux/tags/list
+    [ "$status" -eq 0 ]
+    [ $(echo "${lines[-1]}" | jq '.tags[]') = '"latest"' ]
+
+    run docker kill $(docker ps -q)
+}
+
+@test "sync docker image on demand for Kubernetes - docker" {
```
Comment (andaaron) [inline test/blackbox/sync_docker.bats:153]:
```suggestion
@test "run docker with image synced from registry.k8s.io" {
```

I suggest using this name format for the tests below as well.

File: test/blackbox/sync_docker.bats [hunk 2]
```diff
@@ -80,3 +84,125 @@ function teardown_file() {
     [ "$status" -eq 0 ]
     [ $(echo "${lines[-1]}" | jq '.tags[]') = '"latest"' ]
 }
+
+@test "sync image on demand for kubernetes" {
```
Comment (andaaron) [inline test/blackbox/sync_docker.bats:88]:
```suggestion
@test "sync image on demand from registry.k8s.io" {
```
I suggesting using this naming conventions for other similar tests.

File: test/blackbox/sync_docker.bats [hunk 3]
```diff
@@ -80,3 +84,125 @@ function teardown_file() {
...
+    run skopeo copy docker://127.0.0.1:8090/kube-apiserver-amd64:v1.10.0 oci:${TEST_DATA_DIR} --src-tls-verify=false
+    [ "$status" -eq 0 ]
+
+    run curl http://127.0.0.1:8090/v2/_catalog
+    [ "$status" -eq 0 ]
+    [ $(echo "${lines[-1]}" | jq '.repositories[1]') = '"kube-apiserver-amd64"' ]
```
Comment (andaaron) [inline test/blackbox/sync_docker.bats:94]: '.repositories[1]' seems to be arbitrary. why `1`? How do we make sure these tests don't fail?

Can we do something like https://cloudaffaire.com/faq/how-to-check-if-a-certain-key-or-value-exists-in-json-data-using-jq/ ?
Or something like https://stackoverflow.com/questions/18592173/select-objects-based-on-value-of-variable-in-object-using-jq ?

**Review by andaaron — DISMISSED** [review]
Looks good from my side, @peusebiu please review

File: test/blackbox/sync_docker.bats [hunk 4]
```diff
@@ -80,3 +84,125 @@ function teardown_file() {
...
+    [ "$status" -eq 0 ]
+    [ $(echo "${lines[-1]}" | jq '.tags[]') = '"latest"' ]
+}
+
+@test "run docker with image synced from docker.io/library" {
+    local zot_root_dir=${BATS_FILE_TMPDIR}/zot 
+    run rm -rf ${zot_root_dir}
+    [ "$status" -eq 0 ]
+    
+    run docker run -d 127.0.0.1:8090/archlinux:latest
```
Comment (peusebiu) [inline test/blackbox/sync_docker.bats:141]: also check the status of docker run here

File: test/blackbox/sync_docker.bats [hunk 5]
```diff
@@ -30,7 +30,11 @@ function setup_file() {
             "registries": [
                 {
                     "urls": [
-                        "https://docker.io/library"
+                        "https://docker.io/library",
```
Comment (peusebiu) [inline test/blackbox/sync_docker.bats:33]: nice, having all registries in the same registryConfig.

**Review by peusebiu — APPROVED** [review]

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**